File: Algorithms/Easy/203.Remove-Linked-List-Elements/RemoveLinkedListElements.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummy = ListNode(0, next=head)
        curr = dummy

        while curr and curr.next:
            if curr.next.val == val:
                curr.next = curr.next.next
            else:
                curr = curr.next

        return dummy.next

File: Algorithms/Easy/203.Remove-Linked-List-Elements/RemoveLinkedListElements.go
package easy

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(n)
// Space complexity: O(1)
func removeElements(head *ListNode, val int) *ListNode {
	if head == nil {
		return nil
	}

	dummy := &ListNode{Val: -1, Next: head}
	curr := dummy
	for curr != nil && curr.Next != nil {
		if curr.Next.Val == val {
			curr.Next = curr.Next.Next
		} else {
			curr = curr.Next
		}
	}

	return dummy.Next
}

// Time complexity: O(n)
// Space complexity: O(1)
func removeElements_Recursive(head *ListNode, val int) *ListNode {
	if head == nil {
		return nil
	}

	second := removeElements(head.Next, val)
	if head.Val == val {
		return second
	} else {
		head.Next = second
		return head
	}
}
